### PR TITLE
Core api impl

### PR DESCRIPTION
**Pull request by GiGaByTee**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -29,8 +46,24 @@ apply plugin: 'java'
...
+        threadCount 2
+    }
 }
+
+task apiTest(type: Test) {
+    include '**/*ApiTest.class'
+    systemProperties System.getProperties()
+    useTestNG() {
+        parallel 'methods'
+        threadCount 6
```
Comment (v-lohvyniuk) [inline build.gradle]: could we create fallback for this variables, to make it possible to run it with specifying threadCount via system property etc. ?

File: src/main/java/com/cdp/taf/api/AllureLogListener.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.List;
+
+public class AllureLogListener implements ITestListener, IInvokedMethodListener {
+
+    public void onStart(ITestContext iTestContext) {
+        RestAssured.filters(new ApiLogFilter());
+    }
+
+    @Step("FULL TEST LOG")
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/AllureLogListener.java]: suggest you to replace with "Full log"

File: src/main/java/com/cdp/taf/api/ApiLogFilter.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import io.restassured.filter.Filter;
+import io.restassured.filter.FilterContext;
+import io.restassured.response.Response;
+import io.restassured.specification.FilterableRequestSpecification;
+import io.restassured.specification.FilterableResponseSpecification;
+import org.testng.Reporter;
+
+public class ApiLogFilter implements Filter {
+
+    public ThreadLocal<Integer> counter = new ThreadLocal<>();
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/ApiLogFilter.java]: Could we manage thread local variables via spring context ?

File: src/main/java/com/cdp/taf/api/ApiLogFilter.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    public Response filter(FilterableRequestSpecification requestSpec, FilterableResponseSpecification responseSpec, FilterContext filterContext) {
+        Response response = filterContext.next(requestSpec, responseSpec);
+        Reporter.log("================================================================================================");
+        Reporter.log("=========================================REQUEST================================================");
+        Reporter.log(String.format("WITH METHOD %-10s URI: %-50s", requestSpec.getMethod(), requestSpec.getURI()));
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/ApiLogFilter.java]: WITH is redundant

File: src/main/java/com/cdp/taf/api/ApiLogFilter.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        if (null != requestSpec.getBody()) {
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/ApiLogFilter.java]: use Objects.isNull

File: src/main/java/com/cdp/taf/api/ApiLogFilter.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        return String.format("METHOD: %s \nURI: %s\n", requestSpec.getMethod(), requestSpec.getURI());
+    }
+
+    @Attachment(value = "Response")
+    public String logResponse(Response response) {
+        return String.format("RESPONSE STATUS: %d \nBODY: %s", response.getStatusCode(), response.getBody().prettyPrint());
+    }
+
+    private void updateCounter() {
+        if (counter.get() == null) {
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/ApiLogFilter.java]: Objects.isNull

File: src/main/java/com/cdp/taf/api/services/BaseApiService.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package com.cdp.taf.api.services;
+
+import io.restassured.specification.RequestSpecification;
+
+public class BaseApiService {
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/services/BaseApiService.java]: ApiServiceBase

File: src/main/java/com/cdp/taf/api/services/ClientService.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package com.cdp.taf.api.services;
+
+import io.restassured.response.ValidatableResponse;
+import io.restassured.specification.RequestSpecification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ClientService extends BaseApiService{
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/services/ClientService.java]: format

File: src/main/java/com/cdp/taf/core/ApiConfig.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Scope;
+
+import java.util.Collections;
+
+import static io.restassured.RestAssured.given;
+
+@Configuration
+@ComponentScan(basePackages = "com.cdp.taf.api")
+public class ApiConfig {
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/core/ApiConfig.java]: could you make some base SpringCOnfig ?

File: src/test/java/com/cdp/taf/ClientApiTest.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        ValidatableResponse response = tokenService.getToken();
+        response.statusCode(200);
+        response.body("token_type", equalTo("Bearer"));
+    }
+
+    @Test
+    public void getAllClientsPositiveTest() {
+        ValidatableResponse response = clientService.getClients();
+        response.statusCode(200);
+        List list = responseAs(response, List.class);
```
Comment (v-lohvyniuk) [inline src/test/java/com/cdp/taf/ClientApiTest.java]: Please do not use raw list

File: src/test/java/com/cdp/taf/UserApiTest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    @Test(description = "Get all users and verify list size")
+    public void verifyUsersAmountPositiveTest() {
+        ValidatableResponse response = userService.getUsers();
+        response.statusCode(200);
+        assertEquals((responseAs(response, List.class)).size(), 10);
+    }
+
+    @Test(description = "Verify first user has correct values")
+    public void verifyFirstUserPositiveTest() {
+        ValidatableResponse response = userService.getUser(1);
```
Comment (v-lohvyniuk) [inline src/test/java/com/cdp/taf/UserApiTest.java:33]: so where is your precondition ?

File: src/test/java/com/cdp/taf/ClientApiTest.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.List;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.testng.Assert.assertFalse;
+
+@Listeners(AllureLogListener.class)
+@Test(description = "API test for Client service with authorization")
+public class ClientApiTest extends APITestBase {
+
+    @Autowired
```
Comment (v-lohvyniuk) [inline src/test/java/com/cdp/taf/ClientApiTest.java:21]: can you find the way not to use @Lazy in any place it is autowiring ? Is it possible to use it on class level ?

File: src/main/java/com/cdp/taf/api/dto/TokenDTO.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
+package com.cdp.taf.api.dto;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlRootElement;
+
+@XmlRootElement
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/dto/TokenDTO.java]: suggest you to move this into configuration to use it for all classes

File: build.gradle [hunk 14]
```diff
@@ -1,5 +1,15 @@
+buildscript {
+    repositories {
```
Comment (GiGaByTee) [inline build.gradle]: a

File: build.gradle (hunk 1 above)
Comment (GiGaByTee) [reply build.gradle]: yes, rewritten

File: src/main/java/com/cdp/taf/api/AllureLogListener.java (hunk 2 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/AllureLogListener.java]: done

File: src/main/java/com/cdp/taf/api/ApiLogFilter.java (hunk 6 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/ApiLogFilter.java]: fix

File: src/main/java/com/cdp/taf/api/ApiLogFilter.java (hunk 3 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/ApiLogFilter.java]: I tried, but it doesnt work for me. Moved thread local to separate class

File: src/main/java/com/cdp/taf/api/ApiLogFilter.java (hunk 4 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/ApiLogFilter.java]: agree. Removed

File: src/main/java/com/cdp/taf/api/ApiLogFilter.java (hunk 5 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/ApiLogFilter.java]: used nonNull

File: src/main/java/com/cdp/taf/api/services/BaseApiService.java (hunk 7 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/services/BaseApiService.java]: done

File: src/main/java/com/cdp/taf/api/services/ClientService.java (hunk 8 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/services/ClientService.java]: done

File: src/test/java/com/cdp/taf/ClientApiTest.java (hunk 10 above)
Comment (GiGaByTee) [reply src/test/java/com/cdp/taf/ClientApiTest.java]: updated

File: src/main/java/com/cdp/taf/core/ApiConfig.java (hunk 9 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/core/ApiConfig.java]: done

File: src/test/java/com/cdp/taf/ClientApiTest.java (hunk 12 above)
Comment (GiGaByTee) [reply src/test/java/com/cdp/taf/ClientApiTest.java:21]: refactored not to use Lazy in Test classes

File: src/main/java/com/cdp/taf/api/dto/TokenDTO.java (hunk 13 above)
Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/dto/TokenDTO.java]: done

File: src/main/java/com/cdp/taf/api/Counter.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
+package com.cdp.taf.api;
+
+import java.util.Objects;
+
+public class Counter {
+
+    static ThreadLocal<Integer> counter = new ThreadLocal<>();
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/Counter.java]: use with initial

Comment (GiGaByTee) [reply src/main/java/com/cdp/taf/api/Counter.java]: done

File: src/main/java/com/cdp/taf/api/services/ClientService.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+import io.restassured.response.ValidatableResponse;
+import io.restassured.specification.RequestSpecification;
+import org.springframework.stereotype.Component;
+
+import static io.restassured.RestAssured.given;
+
+@Component
+public class ClientService extends ApiServiceBase {
+
+    public ClientService(RequestSpecification request) {
```
Comment (v-lohvyniuk) [inline src/main/java/com/cdp/taf/api/services/ClientService.java]: rename request to spec

**Review by v-lohvyniuk — APPROVED** [review]

File: src/main/java/com/cdp/taf/api/services/TokenService.java [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+
+import static io.restassured.RestAssured.given;
+
+@Component
+public class TokenService {
+
+    @Autowired
+    RequestSpecification request;
+
+    private String REQUEST_TOKEN_BODY = "{\"client_id\":\"gGxAPVvtikjhfJh4ou8ISdD7DcO6wFEJ\"," +
```
Comment (MykhailoTsiganko) [inline src/main/java/com/cdp/taf/api/services/TokenService.java:16]: final static

**Pull request merged**